thesis_text/thesis_text.md

```diff
@@ -0,0 +1,551 @@
+
+The second Futamura projection observes that _specializer(specializer,interpreter) = compiler_ - we generate a tailored specializer that can transform source code into executables.
+
+The third Futamura projection observes that _specializer(specializer,specializer) = compiler-compiler_, resulting in a tool that takes an _interpreter_ and returns a _compiler_.
+
+In this work, we implement an interpreter and use Truffle to perform the first Futamura projection.
+
+### Guards and de-optimizations
+
+For practical partial evaluation, it is valuable to perform speculative optimizations - compiling the code expecting invariants that can be broken during runtime. One common example of such speculation is optimizations of virtual calls: assuming that the method will always be called on objects of a specific type allows replacing the virtual call with a static one and enables a more aggressive specialization. 
+
+Also, it is often useful to make sure some exceptional code paths are never included in the compilation - for example, if dividing by zero should result in an immediate crash of the application with a message being printed out, there is no use in spending time compiling and optimizing the error-message printing code, as it will at max be called once.
+
+For that, Truffle uses guards - statements that when reached by the runtime result in de-optimization. De-optimization is a process of transferring evaluation from the compiled variant of the method back to the interpreter (at the precise point where it was interrupted) and throwing away the already compiled variant, as its assumptions no longer hold.
+
+As an example, here's a pseudo-code of what a single-cache virtual call implementation could look like.
+
+```
+bool cached = false;
+bool invariant = true;
+type expectedType = null;
+funcptr cache = null;
+
+exec(obj, method)
+{
+    if(!cached)
+    {
+        cached = true;
+        expectedType = obj.Type;
+        cache = obj.Type.ResolveVirtualFunc(method)
+    }
+
+    if (invariant)
+    {
+        if(obj.type==expectedType)
+        {
+            return cache.call()
+        } else {
+            Deoptimize()
+            invariant = false
+        }
+    } 
+    obj.Type.ResolveVirtualFunc(method).call()
+}
+```
+
+When partially evaluated with `invariant == true`, the resulting code will look like this:
+
+![alt](guard_first.drawio.svg)
+
+As long as the virtual call is effectively static at runtime, we only spend time compiling the actual target function (which can be specialized for the environment) and during invocation only pay the price of a simple equality check.
+
+Once the comparison fails, this version of the compiled method is thrown away, and a generic one is created:
```

Comment:
Mozna by se hodilo vysvetlit v cem je to lepsi nez:
```
if (obj.type == expectedType) return cache.call();
// else no deopt
return obj.Type.ResolveVirtualFunc(method).call();
```
takhle delal .NET profile-guided optimalizace posledne, kdyz jsem se na to dival, protoze jeste neumeli deopt, ale umeli tiered kompilaci